### PR TITLE
Add centralized logger

### LINKED ISSUE #4982: Implement a centralized mechanism for logging
# Description
We need to build a logger feature that can help us to log the errors locally in dev environment and send the exceptions and custom events to Azure AppInsights in prod environment 

Proposal:
We build a useLogger hook which returns a logger object with `.info`, `.debug` and `.error` methods.  The hook itself tries to check what environment its in and changes its internal implementation for that, switching between console logging and sending to app insights.

QUESTIONS:
- How can our logger tell which environment its in? Note that if we use APP_ENV, then UAT will behave just like Prod. Do we add a new LOGGER_ENV env variable?
- How should the logger act on UAT? Logging to Azure somehow seems useful, but should be distinguishable from Prod.

# A/C 

- [ ] Logger service implemented in a single place and offers methods to log errors and custom events that can be re-usable within the app 

- [ ] Ability to log in different levels of details : Info , Debug , Error ( https://sematext.com/blog/logging-levels/ ) 

- [ ] Logging destination should be changed based on the environment 
       Local - console 
       Dev, UAT - log files ? 
       PROD - AppInsights

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #4982 

## 👋 Introduction

This branch implements a centralized logger.

## 🕵️ Details

This PR adds, essentially, three loggers.  
1. A console logger than writes messages to the browser console, like `console.log`
2. An Application Insights logger than registers events in Azure
3. A stack logger that combines loggers, [like Laravel has](https://laravel.com/docs/9.x/logging#building-log-stacks)

The stack logger combines the first two and is used as the default logger.

This PR also adds environment configuration for setting what severity level the message must have to be logged.  These should be set in Azure as well.  If the severity level is not set then the logger will not do anything.

## 🧪 Testing

To assist in triggering logging messages, the commit [testing logging buttons](https://github.com/GCTC-NTGC/gc-digital-talent/commit/7d92697213697b9b37fd3810dd48fb85fb00550c) has a set of buttons you can drop onto the homepage.

If you would like to log error messages to Application Insights, you can ask Brinda or I for the connection string.  Or copy it directly from Azure.

1. Add the test logging buttons to code
2. Add the Application Insights connection string to the talentsearch config.ejs file.
3. Copy the new .env variables from the example files (or rerun setup.sh)
4. Rebuild the code
5. Visit the homepage and click all the buttons
6. Observe messages logged to the console
7. Observe messages logged to Application Insights (after 2-3 minutes)

## 📸 Screenshot
image (8)
image

## :truck: Deployment Instructions
Then `LOG_CONSOLE_LEVEL` and `LOG_APPLICATIONINSIGHTS_LEVEL` can be set for each vertical in Azure.  A possible matrix:
|                               | DEV     | UAT   | PROD    |
|-------------------------------|---------|-------|---------|
| LOG_CONSOLE_LEVEL             | debug   | debug | (unset) |
| LOG_APPLICATIONINSIGHTS_LEVEL | (unset) | error | error   |

### CODE REVIEW

File: frontend/common/src/hooks/useLogger.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  notice: (message: string) => {
+    if (levelIncludes(consoleLoggerLevel, SeverityLevel.Notice))
+      console.info(message);
+  },
+  info: (message: string) => {
+    if (levelIncludes(consoleLoggerLevel, SeverityLevel.Info))
+      console.info(message);
+  },
+  debug: (message: string) => {
+    if (levelIncludes(consoleLoggerLevel, SeverityLevel.Debug))
```
Comment (brindasasi) [inline frontend/common/src/hooks/useLogger.ts:51]: I couldn't get this debug message in console

File: frontend/common/src/components/ClientProvider/ClientProvider.tsx [hunk 2]
```diff
@@ -176,8 +182,13 @@ const ClientProvider: React.FC<{
                   toastId: "rate-limit", // limits toasts for rate limit to one.
                 });
 
-              // eslint-disable-next-line no-console
-              console.error(error);
+              logger.error(
```
Comment (brindasasi) [inline frontend/common/src/components/ClientProvider/ClientProvider.tsx:185]: omg.. this is getting real! 😅👏

File: frontend/common/src/hooks/useLogger.ts (hunk 1 above)
Comment (petertgiles) [reply frontend/common/src/hooks/useLogger.ts:51]: Hmm, weird.  The console debug level works OK for me.
image
Could you check in your .env file that LOG_CONSOLE_LEVEL is set to debug and it is passed in your config.js file?
image
image

Comment (brindasasi) [reply frontend/common/src/hooks/useLogger.ts:51]: yeah ..thats what I have.  But only debug button didn't log. 
Screen Shot 2023-01-11 at 10 35 11 AM

Comment (brindasasi) [reply frontend/common/src/hooks/useLogger.ts:51]: oh .. I found the issue .. its my side .. I had settings on my logging level on my console 
Screen Shot 2023-01-11 at 10 45 24 AM

Comment (brindasasi) [reply frontend/common/src/hooks/useLogger.ts:51]: I needed to tick that verbose to see this debug message ..

**Review by brindasasi — APPROVED** [review]
Looks good to me! Lets get this in 🥳

**Pull request merged**